CppCoreGuidelines.md

```diff
@@ -6081,42 +6086,45 @@ If `x = x` changes the value of `x`, people will be surprised and bad errors may
 
     Foo& Foo::operator=(Foo&& a) noexcept  // OK, but there is a cost
     {
-        if (this == &a) return *this;  // this line is redundant
-        s = std::move(a.s);
-        i = a.i;
+        if (this != &a) {
+            s = std::move(a.s);
+            i = a.i;
+        }
         return *this;
     }
 
-The one-in-a-million argument against `if (this == &a) return *this;` tests from the discussion of [self-assignment](#Rc-copy-self) is even more relevant for self-move.
-
-##### Note
-
-There is no known general way of avoiding an `if (this == &a) return *this;` test for a move assignment and still get a correct answer (i.e., after `x = x` the value of `x` is unchanged).
+Here, the test `(this != &a)` is not redundant, because the standard library does not guarantee the result
+of `s = std::move(a.s)` when `s` and `a.s` are the same object. In practice, both libc++ and libstdc++
+will set `s` to the empty string. So, if you want `foo = std::move(foo)` to be a no-op, you must either write
+special-case code to handle the self-assignment case, or use the copy-and-swap idiom.
 
-##### Note
+Prefer to use the copy-and-swap idiom:
 
-The ISO standard guarantees only a "valid but unspecified" state for the standard-library containers. Apparently this has not been a problem in about 10 years of experimental and production use. Please contact the editors if you find a counter example. The rule here is more caution and insists on complete safety.
+    Foo& Foo::operator=(Foo&& a) noexcept  // OK, simplest and best
+    {
+        Foo copy(std::move(a));
+        this->swap(copy);
+        return *this;
+    }
 
 ##### Example
 
-Here is a way to move a pointer without a test (imagine it as code in the implementation a move assignment):
+Here is a way to move a pointer without a test (imagine it as code in the implementation of move assignment):
 
     // move from other.ptr to this->ptr
-    T* temp = other.ptr;
-    other.ptr = nullptr;
+    T* temp = std::exchange(other.ptr, nullptr);
     delete ptr;
     ptr = temp;
```

Comment:
I think the point of the temp variable in the original code was so that you could remember the value of other.ptr after you set it to null. Since std::exchange seems to be designed to do that for you then if you're using std::exchange wouldn't you just write this as
```
delete ptr;
ptr = std::exchange(other.ptr, nullptr);
```
or possibly even
```delete std::exchange(ptr, std::exchange(other.ptr, nullptr));``` ?

Feels like at the moment readability is suffering without any real gain. The same can probably be said of my second example though.

Reply:
This snippet is supposed to keep working even if `&ptr == &other.ptr`, so your two-liner wouldn't work.
Your one-liner seems like it would work fine, though! Totally unreadable, but it does work even when `&ptr == &other.ptr` as far as I can tell.

However, this entire "Note" seems to be just someone showing off their cleverness; it's not code you'd ever use in real life, given that copy-and-swap already exists. So I certainly wouldn't object to just removing the entire "Note."

Reply:
good point about the two liner. Pure coincidence that the one liner works. I agree the one liner is pretty unreadable. If I were reviewing this code in production I'd be advocating for a comment explaining why it's so sensitive (i.e. easy to get wrong). I have no opinion on whether the example is required.